components/server/src/server.rs

```diff
@@ -1466,11 +1400,12 @@ impl TiKVServer<RaftLogEngine> {
                 self.config.storage.block_cache.shared,
```

Comment:
I'd like to make it the same as raft engine, `engines.kv.register_config()`

Reply:
I'm afraid it's impossible unless a similar `ConfiguredKvEngine` trait is definied. But it's just strange to do so as the only implementation is for RocksEngine, and only one method in the trait is `register_config`. Maybe it should be done when KvEngine is abstracted over RocksEngine, PanicsEngine and others.

Reply:
let's revert the `register_config` change, and decuce if it is rocks engine by `config.raft_engine.enable` instead of `as_rocksb_engine`. Then `ConfiguredXXX` is not needed any more. You can simply unify `RaftEngine` and `KvEngine` in the builder pattern then.

Reply:
I don't get it. How can you make the compiler know it's `RocksEngine` without using `as_rocksdb_engine`?

Reply:
if `self.config.raft_engine.enable` 🤣

Reply:
`enable` is a runtime check, the type is inferred at compile time. It means you need to cast a generic to specific time at runtime.

Reply:
but here doesn't need static check, `if raft.as_rocksdb_engine().is_some()` is a runtime check too.

Reply:
I think there is some misunderstanding here. The point is not about runtime check, is about how to make compiler accept a generic and treat it as RocksEngine. This PR achieves this via`as_rocksdb_engine`, so compiler know how to convert a generic to a specific type. With `self.config.raft_engine.enable`, the generic is still a generic, it can't become `RocksEngine` all of a sudden.

It will be better explained by code samples to prove your point.

Reply:
Oh, you mean `EnginesResourceInfo` needs a `RocksEngine`, but what we have is `RaftEngine`.